Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,589 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny {
+  public class PreTypeInferenceState {
+    public readonly PreTypeResolver PreTypeResolver;
+    private readonly DafnyOptions options;
+
+    private List<SubtypeConstraint> unnormalizedSubtypeConstraints = new();
+    private List<EqualityConstraint> equalityConstraints = new();
+    private List<Func<bool>> guardedConstraints = new();
+    private readonly List<Advice> defaultAdvice = new();
+    private List<System.Action> confirmations = new();
+
+    public PreTypeInferenceState(PreTypeResolver preTypeResolver) {
+      this.PreTypeResolver = preTypeResolver;
+      this.options = preTypeResolver.resolver.Options;
+    }
+
+    // ---------------------------------------- Pre-type inference state ----------------------------------------
+
+    /// <summary>
+    /// Solves or simplifies as many type constraints as possible.
+    /// If "allowDecisions" is "false", then no decisions, only determined inferences, are made; this mode is
+    /// appropriate for the partial solving that's done before a member lookup.
+    /// Returns "true" if anything changed (that is, if any of the constraints in the type-inference state
+    /// caused a change some pre-type proxy).
+    /// </summary>
+    void PartiallySolveTypeConstraints(string printableContext = null, bool makeDecisions = false) {
```

Comment:
I would extract the logic related to `makeDecisions` to an outer method. Having boolean parameters that change behavior is a smell, because it ends up with calls like `CanDoAnything(true,true,true,false,false,true)`

```c#
PartiallySettleTypeConstraints(string printableContext = null) {
  bool anythingChanged = false;
  do {
    if (TryResolveTypeProxiesUsingKnownBounds(true)) {
      // something changed, so do another round of Apply... calls below
    } else if (TryResolveTypeProxiesUsingKnownBounds(false)) {
       // something changed, so do another round of Apply... calls below
    } else {
      return;
     }
    anythingChanged = PartiallySolveTypeConstraints(printableContext);
  } while(anythingChanged);
}

bool PartiallySolveTypeConstraints(string printableContext = null) {
  // same as before but without the `if(makeDecisions)` statement, and returning if anythingChanged in any of the loops
}

```

Reply:
So, you're suggesting that the 3 applications that would go into `PartiallySolveTypeConstraints` would always run to a fixpoint before attempting another decision? In other words, there would still be a loop inside `PartiallySolveTypeConstraints`, right? I don't know if that behavior would compare with the current behavior, but I think the code is clearer the way it is.

I also don't agree that parameterized behavior of methods is in general a code smell.

Reply:
> So, you're suggesting that the 3 applications that would go into PartiallySolveTypeConstraints would always run to a fixpoint before attempting another decision?

Yes. I don't understand the process well enough to understand the implication. Could fixpointing before making additional decisions lead to better decisions?